### ISSUE TITLE
Can't get turbodbc to work in AWS Lambda python script

### ISSUE THREAD

**Issue by smaca69** [blue-yonder/turbodbc#186]
I'm trying to run a python script within AWS Lambda which uses turbodbc and cannot get it to execute successfully. I keep on getting the following error in Lambda:

No module named 'turbodbc_intern'

I have included the turbodbc package in the zip file I upload to Lambda but still no success.

Is there something else I need to include in the zip package to upload into Lambda?

**Comment by xhochy** [comment]
How did you upload `turbodbc` to AWS Lambda? What is in your zip? You are probably missing the `turbodbc_*.so` libraries.

**Comment by smaca69** [comment]
Yes, the turbodbc_*.so libraries fixed that error.  Now i'm getting the following error:

Unable to import module 'AATest': libodbc.so.2: cannot open shared object file: No such file or directory

**Comment by xhochy** [comment]
@smaca69 This sounds like you need to read up on how to package Python python environments properly for AWS Lambda. The error is probably related to turbodbc and the database driver but the actual root cause is that you are only uploading selective parts of your required Python environment to Lambda. This tracker is probably the wrong place to get help on that.

**Comment by smaca69** [comment]
I've packaged Python as per AWS documentation but am still getting the following:

Unable to import module 'AATest': /usr/lib64/libstdc++.so.6: version `CXXABI_1.3.8' not found (required by /var/task/turbodbc_intern.cpython-36m-x86_64-linux-gnu.so)

**Comment by MathMagique** [comment]
Sorry, no experience with AWS Lambda here, but the error message indicates that the system you have compiled turbodbc on and the Linux image Amazon uses to provide lambda support different versions of the C++ standard library. To work around, you may need to compile turbodbc on a system that is as close to AWS Lambda as possible. Check out [this tutorial](http://www.perrygeo.com/running-python-with-compiled-code-on-aws-lambda.html), it explains how to do this for another library.

**Comment by bsnithin44** [comment]
Hi @smaca69 

Were you able to solve the problem?. I am having the same issue. Any help would be appreciated.

**Comment by smaca69** [comment]
Hi,

Yes, what are the problems you are having?

Cheers
Sam



Friday, 14 June 2019, 9:15 pm +1000 from notifications@github.com  <notifications@github.com>:

**Comment by bsnithin44** [comment]
I am facing No module named 'turbodbc_intern' error now.

I am a little new to all of this can you tell me where can i find  turbodbc_*.so libraries and should I be adding them directly to the zip file?
What should I do next?

Thanks
Nithin

**Comment by smaca69** [comment]
Have you setup the Lambda environment variables?



Friday, 14 June 2019, 10:00 pm +1000 from notifications@github.com  <notifications@github.com>:

**Comment by bsnithin44** [comment]
No, what am I supposed to do?

**Comment by smaca69** [comment]
These links should be able to help you. If you're still stuck, let me know

https://docs.aws.amazon.com/lambda/latest/dg/lambda-python-how-to-create-deployment-package.html
http://www.perrygeo.com/running-python-with-compiled-code-on-aws-lambda.html

## IMPORTANT - Especially LD_LIBRARY_PATH variable
https://docs.aws.amazon.com/lambda/latest/dg/lambda-environment-variables.html

**Comment by alexanderluiscampino** [comment]
Hi @smaca69,

Is there possibility that you can give some guidance on this? I always get the error of the turbodbc_intern not found.

I followed the links you provided. Do you use the subshell command as they do to run your business logic code with the LD_LIBRARY_PATH set up as you call your python script?

Or can you run your business code within the lambda handler with turbodbc as a imported library?

**Comment by smaca69** [comment]
Hi Alexandre,

No, I dont use the subshell command, just create the Lambda function and
add the following Environment Variable to the Lambda function:

LD_LIBRARY_PATH=/lib64:/usr/lib64:$LAMBDA_RUNTIME_DIR:$LAMBDA_RUNTIME_DIR/lib:$LAMBDA_TASK_ROOT:$LAMBDA_TASK_ROOT/lib:/opt/lib

Append your libraries to the END of the above path E.g:
(/var/task/microsoft/msodbcsql/lib64:/var/task/microsoft/msodbcsql17/lib64:/var/task/lib:/var/task/share:/var/task/include)
e.g.
LD_LIBRARY_PATH=/lib64:/usr/lib64:$LAMBDA_RUNTIME_DIR:$LAMBDA_RUNTIME_DIR/lib:$LAMBDA_TASK_ROOT:$LAMBDA_TASK_ROOT/lib:/opt/lib:/var/task/microsoft/msodbcsql/lib64:/var/task/microsoft/msodbcsql17/lib64:/var/task/lib:/var/task/share:/var/task/include

Hope that helps

Cheers
Sam
